File: app/src/main/java/com/andrewthomas/hackathonmaze/MazePiece.java
package com.andrewthomas.hackathonmaze;

import java.util.Random;

public class MazePiece {
    private Random random = new Random();
    private int locationX, locationY;
    private int tileSize;
    private Directions direction = Directions.NONE;

    public MazePiece(int locationX, int locationY, int tileSize) {
        this.locationX = locationX;
        this.locationY = locationY;
        this.tileSize = tileSize;
    }

    public boolean isOnSameLocationAs(MazePiece otherPiece) {
        return this.locationX == otherPiece.getLocationX() && this.locationY == otherPiece.getLocationY();
    }

    public void generateNewFinishLocation(int mazeSizeX, int mazeSizeY) {
        locationX = random.nextInt(mazeSizeX / 2) + mazeSizeX / 4;
        locationY = random.nextInt(mazeSizeY / 2) + mazeSizeY / 4;
    }

    public boolean move(MazeGenerator maze, Directions direction) {
        if(direction == Directions.WEST) {
            if(locationX != 0 && !maze.getMazeTileAt(locationX, locationY).getIsWallPresent(Directions.WEST)) {
                setLocationX(getLocationX() - 1);
                setDirection(Directions.WEST);
                return true;
            }
        }
        if(direction == Directions.EAST) {
            if(locationX != maze.getSizeX() - 1 && !maze.getMazeTileAt(locationX, locationY).getIsWallPresent(Directions.EAST)) {
                setLocationX(getLocationX() + 1);
                setDirection(Directions.EAST);
                return true;
            }
        }
        if(direction == Directions.NORTH) {
            if(locationY != 0 && !maze.getMazeTileAt(locationX, locationY).getIsWallPresent(Directions.NORTH)) {
                setLocationY(getLocationY() - 1);
                setDirection(Directions.NORTH);
                return true;
            }
        }
        if(direction == Directions.SOUTH) {
            if(locationY != maze.getSizeY() - 1 && !maze.getMazeTileAt(locationX, locationY).getIsWallPresent(Directions.SOUTH)) {
                setLocationY(getLocationY() + 1);
                setDirection(Directions.SOUTH);
                return true;
            }
        }
        return false;
    }

    public int convertCoordinateToLocation(double coordinate, int tileSize) {
        return (int) ((coordinate + tileSize / 2) / tileSize) - 1;
    }

    public void setLocationX(int locationX) {
        this.locationX = locationX;
    }

    public void setLocationY(int locationY) {
        this.locationY = locationY;
    }

    public void setDirection(Directions direction) {
        this.direction = direction;
    }

    public int getLocationX() {
        return locationX;
    }

    public int getLocationY() {
        return locationY;
    }

    public int getCoordinateX() {
        return (locationX + 1) * tileSize;
    }

    public int getCoordinateY() {
        return (locationY + 1) * tileSize;
    }

    public Directions getDirection() {
        return direction;
    }
}

File: app/src/main/com/andrewthomas/hackathonmaze/MazePiece.kt
package com.andrewthomas.hackathonmaze

import java.util.Random

class MazePiece(var locationX: Int, var locationY: Int, val tileSize: Int) {
    private val random = Random()
    var coordinateX: Int = 0
    var coordinateY: Int = 0
    var oldCoordinateX: Int = 0
        private set
    var oldCoordinateY: Int = 0
        private set
    var direction = Directions.NONE

    init {
        this.coordinateX = (locationX + 1) * tileSize
        this.coordinateY = (locationY + 1) * tileSize
    }

    fun isOnSameLocationAs(otherPiece: MazePiece): Boolean {
        return this.locationX == otherPiece.locationX && this.locationY == otherPiece.locationY
    }

    fun isOnSameLocationAs(mazeTile: MazeTile): Boolean {
        return this.locationX == mazeTile.locationX && this.locationY == mazeTile.locationY
    }

    fun generateNewFinishLocation(mazeSizeX: Int, mazeSizeY: Int) {
        locationX = random.nextInt(mazeSizeX / 2) + mazeSizeX / 4
        locationY = random.nextInt(mazeSizeY / 2) + mazeSizeY / 4
        coordinateX = (locationX + 1) * tileSize
        coordinateY = (locationY + 1) * tileSize
    }

    fun move(maze: MazeGenerator, direction: Directions): Boolean {
        oldCoordinateX = coordinateX
        oldCoordinateY = coordinateY
        if (direction === Directions.WEST) {
            if (locationX != 0 && !maze.getMazeTileAt(locationX, locationY).getIsWallPresent(Directions.WEST)) {
                locationX -= 1
                return true
            }
        }
        if (direction === Directions.EAST) {
            if (locationX != maze.sizeX - 1 && !maze.getMazeTileAt(locationX, locationY).getIsWallPresent(Directions.EAST)) {
                locationX += 1
                return true
            }
        }
        if (direction === Directions.NORTH) {
            if (locationY != 0 && !maze.getMazeTileAt(locationX, locationY).getIsWallPresent(Directions.NORTH)) {
                locationY -= 1
                return true
            }
        }
        if (direction === Directions.SOUTH) {
            if (locationY != maze.sizeY - 1 && !maze.getMazeTileAt(locationX, locationY).getIsWallPresent(Directions.SOUTH)) {
                locationY += 1
                return true
            }
        }
        return false
    }

    fun move(newLocationX: Int, newLocationY: Int): Boolean {
        oldCoordinateX = coordinateX
        oldCoordinateY = coordinateY
        if (locationX == newLocationX) {
            if (locationY < newLocationY) {
                direction = Directions.SOUTH
            } else {
                direction = Directions.NORTH
            }
        } else {
            if (locationX < newLocationX) {
                direction = Directions.EAST
            } else {
                direction = Directions.WEST
            }
        }
        locationX = newLocationX
        locationY = newLocationY
        return true
    }
}
